# Importing the required libraries and dataset.

In [3]:
import keras
keras.__version__

'2.8.0'

In [4]:
from keras.datasets import imdb

(train_data, train_labels), (test_data, test_labels) = imdb.load_data(num_words=10000)

In [5]:
max([max(sequence) for sequence in train_data])

9999

In [6]:
import numpy as np

def vectorize_sequences(sequences, dimension=10000):
    # Create an all-zero matrix of shape (len(sequences), dimension)
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1.  # set specific indices of results[i] to 1s
    return results

# Our vectorized training data
x_train = vectorize_sequences(train_data)
# Our vectorized test data
x_test = vectorize_sequences(test_data)

In [7]:
# Our vectorized labels
y_train = np.asarray(train_labels).astype('float32')
y_test = np.asarray(test_labels).astype('float32')

# Training the dataset with 3 different configuration

In [8]:
from keras import models
from keras import layers

#Neural Network with the same parameters from what the SME did.
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)

Epoch 1/4
49/49 [==============================] - 2s 26ms/step - loss: 0.4292 - accuracy: 0.8251
Epoch 2/4
49/49 [==============================] - 1s 25ms/step - loss: 0.2501 - accuracy: 0.9114
Epoch 3/4
49/49 [==============================] - 1s 24ms/step - loss: 0.1987 - accuracy: 0.9287
Epoch 4/4
782/782 [==============================] - 2s 2ms/step - loss: 0.2936 - accuracy: 0.8832


In [9]:
results

[0.29359251260757446, 0.8831999897956848]

In [10]:
# Using 64 hidden units, mse for the loss function and tahn for the activation function
model = models.Sequential()
model.add(layers.Dense(64, activation='tanh', input_shape=(10000,)))
model.add(layers.Dense(64, activation='tanh'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='MeanSquaredError',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)

Epoch 1/4
49/49 [==============================] - 2s 40ms/step - loss: 0.1308 - accuracy: 0.8131
Epoch 2/4
49/49 [==============================] - 2s 39ms/step - loss: 0.0653 - accuracy: 0.9148
Epoch 3/4
49/49 [==============================] - 2s 39ms/step - loss: 0.0521 - accuracy: 0.9319
Epoch 4/4
782/782 [==============================] - 2s 2ms/step - loss: 0.1032 - accuracy: 0.8674


In [11]:
results

[0.10319385677576065, 0.86735999584198]

In [12]:
# Using 32 hidden units, mse for the loss function and tahn for the activation function
model = models.Sequential()
model.add(layers.Dense(32, activation='tanh', input_shape=(10000,)))
model.add(layers.Dense(32, activation='tanh'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='MeanSquaredError',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)

Epoch 1/4
49/49 [==============================] - 2s 28ms/step - loss: 0.1273 - accuracy: 0.8252
Epoch 2/4
49/49 [==============================] - 1s 29ms/step - loss: 0.0641 - accuracy: 0.9180
Epoch 3/4
49/49 [==============================] - 1s 28ms/step - loss: 0.0497 - accuracy: 0.9365
Epoch 4/4
782/782 [==============================] - 2s 2ms/step - loss: 0.0961 - accuracy: 0.8764


In [13]:
results

[0.09612084180116653, 0.8763999938964844]

**Conclusion: The first setup (the one that SME did) has the highest accuracy compared to the 2 setups with another parameters.  
1st: 88.3%  
2nd: 86.7%  
3rd: 87.6  
Recommendation: Perhaps we can check the training and validation loss to check with the optimal number of epoch and used the 1st setup in hope for a higher accuracy**
